src/samplers/abc.jl

```diff
@@ -0,0 +1,146 @@
+#################### Approximate Bayesian Computing ####################
+
+#################### Types and Constructors ####################
+
+type ABCTune
+  Told::Matrix{Float64}
+  epsilon::Vector{Float64}
+  terminals::Vector{Symbol}
+end
+
+ABCTune() = ABCTune(Matrix{Float64}(), Vector{Float64}(), Vector{Symbol}())
+
+function ABC(params::Vector{Symbol}, sigma::Real, summarize::Vector{Function},
+             target::Real; rho::Function = (x, y) -> sqrt(sumabs2(x - y)),
+             s::Integer = 1, n::Integer = 50, kernel::Symbol = :uniform)
+
+  kernel in [:uniform, :epanechnikov, :gaussian] ||
+    throw(ArgumentError(
+      "kernel must be one of :uniform, :epanechnikov, or :gaussian"
+    ))
+
+  samplerfx = function(model::Model, block::Integer)
+    tunepar = tune(model, block)
+
+    ## previous value
+    x = unlist(model, block, true)
+    logfx = mapreduce(p -> logpdf(model[p], true), +, params)
+    
+
+    T0 = NaN
+    Told = NaN
+    Tnew = NaN
+
+    ## Initialize tune
+    if model.iter == 1
+      terminals = keys(model, :output)
+      nt = length(terminals)
+
+      ## data
+      dvec = Array(Array{Float64, 1}, nt)
+      for d in 1:nt
+        dvec[d] = unlist(model, [terminals[d]])
+      end
+
+      ## summary stat of data
+      T0vec = Array(Array{Float64, 1}, nt)
+      for d in 1:nt
+        T0vec[d] = vcat(map(f -> f(dvec[d]), summarize)...)
+      end
+      T0 = vcat(T0vec...)
+
+      m = length(T0)
+
+      Told = zeros(m, s)
+      eps0 = zeros(s)
+      for i in 1:s
+        Toldvec = Array(Array{Float64, 1}, nt)
+        for d in 1:nt
+          x0 = unlist(model[terminals[d]], rand(model[terminals[d]]))
+          Toldvec[d] = vcat(map(f -> f(x0), summarize)...)
+        end
+        Told[:, i] = vcat(Toldvec...)
+        eps0[i] = rho(Told[:, i], T0)
+      end
+      tunepar.Told = Told
+      tunepar.epsilon = eps0
+      tunepar.terminals = terminals
+    else
+      nt = length(tunepar.terminals)
+      ## data
+      dvec = Array(Array{Float64, 1}, nt)
+      for d in 1:nt
+        dvec[d] = unlist(model, [terminals[d]])
+      end
```

Comment:
Should this line reference ``tunepar.terminals`` instead of ``terminals``?

Also, might consider simplifying this and the analogous code in the if condition above to something like

```julia
dvec = map(node -> unlist(model, [node]), tunepar.terminals)
```